### DISCUSSION TITLE
Proper way to programmatically create pages?

### DISCUSSION THREAD

**Discussion by mjmare** [wagtail/wagtail#11778] [Q&A]
Hi all, 

I have created a Django management command that creates the initial Wagtail site structure. This works reasonably well, but I have encountered a problem: Although the pages are shown as "Live" in the WT UI, they are not really. Pageurls render as None, so links are broken. It can be easily worked around by manually publishing the pages again. Then everything works as normal. 
I'd really like to know what I'm doing wrong. 

This is the relevant code:

```python
    def handle(self, *args, **options):
        site = Site.objects.get(is_default_site=True)
        root = Page.get_first_root_node()
        old_home = root.get_first_child()  # don't throw away yet because this will remove the site as well!

        new_home = HomePage(title="Testwijzer home")
        root.add_child(instance=new_home)

        site.root_page = new_home
        site.site_name = 'Testwijzer'
        site.save()

        # now we can remove the old home page
        old_home.delete()

        # change the home page slug:
        new_home.slug = 'home'
        new_home.save_revision().publish()

        self.stdout.write(self.style.SUCCESS('Changed home page successfully.'))

        # Check if AspectIndex page exists; if not, create it
        aspect_index_page = AspectIndexPage.objects.first()
        if not aspect_index_page:
            aspect_index_page = AspectIndexPage(title="Aspect Index",
                                                slug='aspect-index',
                                                body='Overzicht van alle aspecten')
            new_home.add_child(instance=aspect_index_page)
            self.stdout.write(self.style.SUCCESS('AspectIndexPage page added successfully.'))
        else:
            self.stdout.write(self.style.WARNING('AspectIndexPage already exists.'))

```
In this example the Aspect Index Page would exhibit the erroneous behavior.

On the topic of programatically creating objects: AFAICT this is not documented, ssh the code above is what I think is the right way. Especially the relation between Site and Root and pages is somewhat nebulous to me.
If someone could share some example code on the proper way to do this, that would be wonderful.

**Comment by ababic** [comment]
@mjmare as someone who as overseen a few large-scale content migrations from other CMSs to Wagtail, I have the following advice:

1. Forget about generating revisions and publishing them. Keep it simple by writing directly to the real tables only.
2. Use `live=True` in your values to create a live/published page. It will also pay to familiarise yourself with other 'core' model fields on the page model... There's a lot of handy stuff in there.
3. Using `existing_page.add_child(instance=new_page)` is definitely the way to go! You can build up some big trees that way if necessary. Just be aware that the action is NOT atomic, so I'd recommend wrapping your code in `django.db.transaction.atomic` for some added protection, if you're operating on a production database.
4. If adding lots of pages to the same part of the tree, you just have to live with the slowness of adding pages one-by-one - using a threadpool will only create headaches. That said, writing to entirely different branches of the tree simultaneously 'should' be fine.
5. If you're adding lots of children to the same page in a loop, in a way that skips over failures to continue with the next one, be sure to use `parent_page.refresh_from_db(["numchild"])` to correct the in-memory value before attempting to create more pages. Otherwise, you'll run into some errors that are difficult to reason about.
6. For simple examples of creating pages programmatically, look no further than Wagtail's test suite. There are literally hundreds of examples in there.

Hope this information helps others!

**Reply by mjmare** [reply]
Very useful info. Thank you!
